### ISSUE TITLE
Alternative syntax suggestions

### ISSUE THREAD

**Issue by KelseyDH** [apneadiving/waterfall#13]
This is an impressive DSL and I really like how it flows. 😉  However, I have to admit I find the concepts of `dam` and `outflows` to be confusing terminology.  I love the railway analogy, but as it applies to water I find a bit confusing. The railway analogy is taught as `success` and `failure` so ideally we would control flow using those concepts.

So when I think of what would be most human readable for handling problems, it would be language like `on_breach`, `on_failure` or `on_error`.

Likewise for assessing state, instead of `outflow` I quite like `context` which is becoming the idiomatic way we see it done in gems like Interactors.

Otherwise I love this DSL! Cheers!

**Comment by apneadiving** [comment]
Hi Kelsey,

thanks for the feedback :)
I can understand all this water analogy does not please all minds!

I would not compare `outflow` with `context` from Interactors. In Interactor the context is some shared object leaking from one class to the other (I wonder what happens if two interactors happen to name variables the same way).
`outflow` would rather be a `local_context`, only attached to one object. It can pass values to another `outflow` upon request, that's it.

waterfall is a mini-gem, 200-ish LOC, so it is pretty much flexible. Here is a bonus if you want to get rid of water-ish syntax ;)

```
class MyInteractor 
  include Waterfall

  alias :on_failure :on_dam
  alias :local_context :outflow
end

class MyService < MyInteractor
   # ...
end
```
Cheers!

**Comment by artsyca** [comment]
The concept is not unlike `Promises` in `Javascript` -- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

```
Promise.all()
Promise.allSettled()
Promise.prototype.catch()
Promise.prototype.finally()
Promise.prototype.then()
Promise.race()
Promise.reject()
Promise.resolve()
```

**Comment by apneadiving** [comment]
@artsyca yes totally.
Yet there is a ruby gem called Promise and I prefer not to take the same naming as code is most of the time synchronous here

**Comment by artsyca** [comment]
No, let's not even go there -- the waterfall metaphor is lovely
